### PR TITLE
Filters enhancements

### PR DESCRIPTION
**Pull request by mohawk2**
Intended to increase the usefulness of filters. Allows them to be configured with parameters under `x-filter`, and also to be configured on OpenAPI operations.

Also adds two filters: `yancy.from_helper` and `yancy.overlay_from_helper`. These are intended for replacing either a single property, or applying to a collection or operation-input hash-ref, with a named helper.

### CODE REVIEW

**Comment by mohawk2** [comment]
Further commits also added output filtering with `x-filter-output` and filter `yancy.wrap`.

File: lib/Yancy/Controller/Yancy/API.pm [hunk 1]
```diff
@@ -193,9 +193,7 @@ sub delete_item {
 # confusing to understand what the problem is
 sub _delete_null_values {
     for my $item ( @_ ) {
-        for my $key ( grep { !defined $item->{ $_ } } keys %$item ) {
-            delete $item->{ $key };
-        }
+        delete $item->{ $_ } for grep !defined $item->{ $_ }, keys %$item;
```
Comment (preaction) [inline lib/Yancy/Controller/Yancy/API.pm]: This sounds like a performance enhancement. How much does it improve performance? Could you add that information to the commit message? (Why this change is being made and what impact it has?)

File: lib/Mojolicious/Plugin/Yancy.pm [hunk 2]
```diff
@@ -391,6 +391,23 @@ executed in that order. E.g.:
         },
     }
 
+You can also configure filters on OpenAPI operations' outputs, this time
```
Comment (preaction) [inline lib/Mojolicious/Plugin/Yancy.pm]: This sounds like it's useful, but it also sounds like it could hide problems better solved by other mechanisms (such as making sure the data is formatted correctly inside the database, or making the editor handle the formats better). What kind of use-cases are you seeing for this feature?

File: lib/Mojolicious/Plugin/Yancy.pm [hunk 3]
```diff
@@ -471,6 +471,18 @@ This configuration will achieve the same as the above with C<last_updated>:
         },
     }
 
+=head4 yancy.wrap
+
+The configured parameters are a list of strings. For each one, the
```
Comment (preaction) [inline lib/Mojolicious/Plugin/Yancy.pm]: This is another one: Could you add a real-life use-case to the docs to help users understand what they might end up using this for?

File: lib/Yancy/Controller/Yancy/API.pm (hunk 1 above)
Comment (mohawk2) [reply lib/Yancy/Controller/Yancy/API.pm]: I'm just rewording the message. I have not benchmarked it, but I have heard many "proper" Perl experts (unlike me) say that is one way to get better performance.

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 2 above)
Comment (mohawk2) [reply lib/Mojolicious/Plugin/Yancy.pm]: Specifically in the "realworld" API, they annoyingly have one- or two-level "objects" that wrap a simple input or output. Output filtering preserves the simplicity of the DB, while conforming to that shape. Since filters can do just about anything, this is only one example! Another might be to add a SHA.

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 3 above)
Comment (mohawk2) [reply lib/Mojolicious/Plugin/Yancy.pm]: As noted above, the example is literally from the "realworld" API.

File: lib/Yancy/Controller/Yancy/API.pm (hunk 1 above)
Comment (mohawk2) [reply lib/Yancy/Controller/Yancy/API.pm]: Force-pushed with updated message.

Comment (mohawk2) [reply lib/Yancy/Controller/Yancy/API.pm]: Here is an investigation from 2006 (https://perlmonks.org/?node_id=536503). My reasoning was that given `_delete_null_values` is such a "hot" bit of code, called for most operations, that even the chance of a bit of extra speed (and for code I find more readable, but you might disagree!) was worth it.

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 3 above)
Comment (mohawk2) [reply lib/Mojolicious/Plugin/Yancy.pm]: I could add to the docs saying it's from the "realworld" thingy if you like?

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 2 above)
Comment (preaction) [reply lib/Mojolicious/Plugin/Yancy.pm]: Hmm... or a "session" could start by submitting a `POST` request to a `/session` API with a token and then, like you mention, the `x-filter-output` could add the ephemeral stuff. Or data from another source (LDAP, cache, etc...). Okay.

I was thinking that you may not know I was planning on having backend hooks (hooks which could then be added by plugins): One for before/after the five methods. This may result in some things being better added to certain layers:

* Filters operate on the controller layer for the web
* Hooks operate on the backend layer for anything

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 3 above)
Comment (preaction) [reply lib/Mojolicious/Plugin/Yancy.pm]: I mean more that "outer" and "inner" are placeholders for real data, and we could just use the real data instead. It's trying to flatten/unflatten data into a relational database, yes? Like I could do:

```
{
    user => {
        name => 'Foo',
        email => 'doug@preaction.me',
    }
}
```

And by doing a `wrap` / `unwrap` pair change it to:

```
{
    name => 'Foo',
    email => 'doug@preaction.me',
}
```

Right?

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 2 above)
Comment (mohawk2) [reply lib/Mojolicious/Plugin/Yancy.pm]: I've been thinking along vaguely similar lines in that for instance it will be good to have a hookable OpenAPI-inference phase. It may be worth having a more generalised mechanism, possibly tied to Mojolicious's "events".

File: lib/Mojolicious/Plugin/Yancy.pm (hunk 3 above)
Comment (mohawk2) [reply lib/Mojolicious/Plugin/Yancy.pm]: Right! And done :-) (see force-pushed update)

**Comment by mohawk2** [comment]
If you feel the issues are all resolved (thanks, your reviews always lead to improvements!), do you feel this ready for merging? I rebased against your latest `master`.

**Comment by preaction** [comment]
Yes, sorry, I knew I forgot something.
